diff --git a/dlls/kernelbase/Makefile.in b/dlls/kernelbase/Makefile.in
index 081e9bfbb64..1c75a638002 100644
--- a/dlls/kernelbase/Makefile.in
+++ b/dlls/kernelbase/Makefile.in
@@ -3,6 +3,7 @@ MODULE    = kernelbase.dll
 IMPORTLIB = kernelbase
 IMPORTS   = uuid ntdll winecrt0
 
+EXTRA_CFLAGS = -Wa,-muse-unaligned-vector-move
 EXTRADLLFLAGS = -nodefaultlibs -nostartfiles
 i386_EXTRADLLFLAGS = -Wl,--image-base,0x7b000000
 x86_64_EXTRADLLFLAGS = -Wl,--image-base,0x174000000
diff --git a/tools/makedep.c b/tools/makedep.c
index 21c94ee5641..5b7a4d478f6 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -185,6 +185,7 @@ struct makefile
     struct strarray include_args;
     struct strarray define_args;
     struct strarray unix_cflags;
+    struct strarray extra_cflags;
     struct strarray programs;
     struct strarray scripts;
     struct strarray imports;
@@ -3227,6 +3228,7 @@ static void output_source_one_arch( struct makefile *make, struct incl_file *sou
         strarray_add( &make->clean_files, obj_name );
 
     if (!source->use_msvcrt) strarray_addall( &cflags, make->unix_cflags );
+    else strarray_addall( &cflags, make->extra_cflags );
     if ((source->file->flags & FLAG_ARM64EC_X64) && !strcmp( archs.str[arch], "arm64ec" ))
     {
         var_cc     = "$(x86_64_CC)";
@@ -3241,6 +3243,7 @@ static void output_source_one_arch( struct makefile *make, struct incl_file *sou
         strarray_addall( &cflags, make->extlib ? extra_cflags_extlib[arch] : extra_cflags[arch] );
     }
 
+
     if (!arch)
     {
         if (source->file->flags & FLAG_C_UNIX)
@@ -4473,6 +4476,7 @@ static void load_sources( struct makefile *make )
     make->include_args = empty_strarray;
     make->define_args = empty_strarray;
     make->unix_cflags = empty_strarray;
+    make->extra_cflags = empty_strarray;
     if (!make->extlib) strarray_add( &make->define_args, "-D__WINESRC__" );
     strarray_add( &make->unix_cflags, "-DWINE_UNIX_LIB" );
 
@@ -4494,6 +4498,7 @@ static void load_sources( struct makefile *make )
     }
     strarray_addall( &make->define_args, get_expanded_make_var_array( make, "EXTRADEFS" ));
     strarray_addall_uniq( &make->unix_cflags, get_expanded_make_var_array( make, "UNIX_CFLAGS" ));
+    strarray_addall_uniq( &make->extra_cflags, get_expanded_make_var_array( make, "EXTRA_CFLAGS" ));
 
     strarray_add( &make->include_args, strmake( "-I%s", obj_dir_path( make, "" )));
     if (make->src_dir)
