diff --git a/dlls/kernelbase/thread.c b/dlls/kernelbase/thread.c
index 12e1777262a..c53092a65dc 100644
--- a/dlls/kernelbase/thread.c
+++ b/dlls/kernelbase/thread.c
@@ -338,6 +338,7 @@ LANGID WINAPI DECLSPEC_HOTPATCH GetThreadUILanguage(void)
 /***********************************************************************
  *	     OpenThread   (kernelbase.@)
  */
+__attribute__((optimize("no-tree-vectorize")))
 HANDLE WINAPI DECLSPEC_HOTPATCH OpenThread( DWORD access, BOOL inherit, DWORD id )
 {
     HANDLE handle;
