diff --git a/dlls/kernelbase/thread.c b/dlls/kernelbase/thread.c
index 12e1777262a..b50224175f3 100644
--- a/dlls/kernelbase/thread.c
+++ b/dlls/kernelbase/thread.c
@@ -344,6 +344,7 @@ HANDLE WINAPI DECLSPEC_HOTPATCH OpenThread( DWORD access, BOOL inherit, DWORD id
     OBJECT_ATTRIBUTES attr;
     CLIENT_ID cid;
 
+    memset(&attr, 0, sizeof(attr));
     attr.Length = sizeof(attr);
     attr.RootDirectory = 0;
     attr.Attributes = inherit ? OBJ_INHERIT : 0;
