diff --git a/libs/vkd3d/Makefile.in b/libs/vkd3d/Makefile.in
index 868f4582380..51007ebc99c 100644
--- a/libs/vkd3d/Makefile.in
+++ b/libs/vkd3d/Makefile.in
@@ -1,4 +1,4 @@
-EXTLIB    = libvkd3d.a
+STATICLIB = libvkd3d.a
 EXTRAINCL = $(VKD3D_PE_CFLAGS) -I$(srcdir)/include/private -I$(srcdir)/libs/vkd3d -I$(srcdir)/libs/vkd3d-shader
 EXTRADEFS = \
 	-DWINE_NO_NAMELESS_EXTENSION \
diff --git a/libs/vkd3d/libs/vkd3d-shader/fx.c b/libs/vkd3d/libs/vkd3d-shader/fx.c
index 8b7fde70db2..89363a8a549 100644
--- a/libs/vkd3d/libs/vkd3d-shader/fx.c
+++ b/libs/vkd3d/libs/vkd3d-shader/fx.c
@@ -19,6 +19,7 @@
  */
 
 #include "hlsl.h"
+#include <sys/types.h>
 
 #define TAG_FX20 0x46580200
 
